### ISSUE TITLE
Performance Problem Sync vs Async

### ISSUE THREAD

**Issue by vinhhrv** [StackExchange/StackExchange.Redis#661]
Use Async Method take 35 second
But when i run with sync method, take 9 seconds
It is problem, response time async very high when high load

```
class Program
    {
        private static IDatabase _redis;

        static Program()
        {
            _redis = ConnectionMultiplexer.Connect("localhost:6379").GetDatabase();
        }

        static void runasync()
        {
            Enumerable.Range(0, 100).ToList().AsParallel().WithDegreeOfParallelism(50).ForAll(a => {
                try
                {
                    var id = Guid.NewGuid().ToString();
                    _redis.StringSetAsync(id, new byte[] { 1 }, TimeSpan.FromSeconds(3)).GetAwaiter().GetResult();
                }
                catch
                {

                }
                finally
                {
                 //   Console.WriteLine($"runed {a}");
                }
            });
        }

        static void runsync()
        {
            Enumerable.Range(0, 100).ToList().AsParallel().WithDegreeOfParallelism(50).ForAll(a => {
                try
                {
                    var id = Guid.NewGuid().ToString();
                    _redis.StringSet(id, new byte[] { 1 }, TimeSpan.FromSeconds(3));
                }
                catch
                {

                }
                finally
                {
                   // Console.WriteLine($"runed {a}");
                }
            });
        }

        static void Main(string[] args)
        {
            Console.WriteLine("Hello World!");

            new loadtest().get();

            var watch = new Stopwatch();
            watch.Start();
            runsync();
            var timeforsync = watch.Elapsed.TotalSeconds;
            watch.Restart();
            runasync();
            var timeforasync = watch.Elapsed.TotalSeconds;

            Console.WriteLine($"Time For Sync {timeforsync} vs async {timeforasync}");
            Console.ReadLine();
        }
    }
```

**Comment by mgravell** [comment]
it looks to me like you're blocking the thread-pool; you start up to 50
async operations, each of which is actually "sync over async" because
you're calling .GetResult(). However, the callback operation *also* needs
use of that queue, and it won't be able to do anything until the thread
pool grows large enough to have a spare thread (because: they're all
waiting and blocking). And the thread pool grows slowly.

Basically, don't do GetAwaiter.GetResult(), Instead, use an actual
async/await; with await, it won't block, so the thread will be available to
service the callback, which in turn will execute the continuation.

A small amount of indirection is required here because of how continuations
work; it is necessary to avoid the risk of a continuation stealing the
reader thread (potentially indefinitely).
